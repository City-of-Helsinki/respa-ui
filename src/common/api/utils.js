import get from 'lodash/get';

import constants from '../../../app/constants/AppConstants';
import store from '../../store';

let authToken;

store.subscribe(() => {
  const state = store.getState();
  authToken = get(state, 'auth.token');
});

function getAuthorizationHeader() {
  return authToken
    ? { Authorization: `JWT ${authToken}` }
    : {};
}

/**
 * Getter for API url.
 * @param endpoint
 * @param params
 * @param trailingSlash This is because Django adds trailing slash into every endpoint.
 * @returns {string}
 */
export const getUrl = (endpoint, params = {}, trailingSlash = true) => {
  const query = Object.keys(params).map(key => `${key}=${params[key]}`).join('&');
  return `${constants.API_URL}/${endpoint}${trailingSlash ? '/' : ''}${query ? `?${query}` : ''}`;
};

/**
 * Getter for default headers.
 * @returns {object}
 */
export const getHeaders = () => ({
  ...constants.REQUIRED_API_HEADERS,
  ...getAuthorizationHeader(),
});
