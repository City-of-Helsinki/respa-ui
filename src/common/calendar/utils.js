export const getDefaultSelectedTimeRange = (reservation) => {
  return reservation ? {
    start: reservation.begin,
    end: reservation.end
  } : null;
};
