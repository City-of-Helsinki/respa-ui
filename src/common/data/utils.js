import get from 'lodash/get';
import values from 'lodash/values';
import pickBy from 'lodash/pickBy';

export const getLocalizedFieldValue = (field, locale, fallback = false) => {
  const localeValue = get(field, locale, null);

  if (localeValue || !fallback) {
    return localeValue;
  }

  return values(field).find(fallbackValue => !!fallbackValue);
};


export const parseData = (data) => {
  return pickBy(data, value => value || value === 0);
};
