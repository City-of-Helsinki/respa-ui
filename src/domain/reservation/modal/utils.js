export const getCancelCategories = (data = [], isAdmin, locale) => {
  return data
    .filter(({ reservation_type: reservationType }) => (
      isAdmin ? true : reservationType === 'own'))
    .map(category => ({
      value: category.id,
      label: category.name[locale || 'fi'],
    }));
};
