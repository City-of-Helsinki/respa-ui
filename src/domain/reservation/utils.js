import moment from 'moment';
import find from 'lodash/find';
import sortBy from 'lodash/sortBy';
import clone from 'lodash/clone';
import tail from 'lodash/tail';
import last from 'lodash/last';

export const combineReservations = (reservations) => {
  if (!reservations || !reservations.length) {
    return [];
  }

  const sorted = sortBy(reservations, 'begin');
  const initialValue = [clone(sorted[0])];

  return tail(sorted).reduce((previous, current) => {
    if (current.begin === last(previous).end) {
      last(previous).end = current.end;
    } else {
      previous.push(clone(current));
    }
    return previous;
  }, initialValue);
};

export const getNextAvailableTime = (reservations, fromMoment = moment()) => {
  const combinedReservations = combineReservations(reservations);
  if (!combinedReservations.length || fromMoment < moment(combinedReservations[0].begin)) {
    return fromMoment;
  }

  const ongoingReservation = find(combinedReservations, reservation => (
    moment(reservation.begin) <= fromMoment && fromMoment < moment(reservation.end)
  ));

  return ongoingReservation ? moment(ongoingReservation.end) : fromMoment;
};

export const getCurrentReservation = (reservations) => {
  const now = moment();
  return find(reservations, (reservation) => {
    return moment(reservation.begin) < now && now < moment(reservation.end);
  });
};
