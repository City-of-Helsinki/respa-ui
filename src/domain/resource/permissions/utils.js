import difference from 'lodash/difference';

import { resourceRoles, resourcePermissionTypes, resourcePermissionsByRole } from './constants';

// Returns the role with the most permissions or null.
export function roleMapper(isAdmin, isManager, isViewer) {
  if (isAdmin) {
    return resourceRoles.UNIT_ADMINISTRATOR;
  }

  if (isManager) {
    return resourceRoles.UNIT_MANAGER;
  }

  if (isViewer) {
    return resourceRoles.UNIT_VIEWER;
  }

  return null;
}

const permissionOptions = Object.values(resourcePermissionTypes);
const allValuesInArray = (subset, superset) => difference(subset, superset).length === 0;

export function hasPermissionForResource(resourceRole, requiredPermissions) {
  // undefined roles don't receive any permissions implicitly
  if (resourceRole === undefined || resourceRole === null) {
    return false;
  }

  if (!Object.values(resourceRoles).includes(resourceRole)) {
    throw Error(`resourceRole needs to be one of: ${Object.values(resourceRoles).join(', ')}.`);
  }

  if (!Array.isArray(requiredPermissions) && typeof requiredPermissions !== 'string') {
    throw Error('requiredPermissions has to be either a string or an array.');
  }

  if (
    Array.isArray(requiredPermissions)
    && !allValuesInArray(requiredPermissions, permissionOptions)
  ) {
    const unknownPermissions = difference(requiredPermissions, permissionOptions);

    // eslint-disable-next-line no-console
    console.warn([
      `Permission asked with unknown permission: ${unknownPermissions.join(', ')}.`,
      `Permission needs to be on of: ${permissionOptions.join(', ')}.`,
    ].join('\n'));
  }

  if (typeof requiredPermissions === 'string' && !permissionOptions.includes(requiredPermissions)) {
    // eslint-disable-next-line no-console
    console.warn([
      `Permission asked with unknown permission: ${requiredPermissions}.`,
      `Permission needs to be on of: ${permissionOptions.join(', ')}.`,
    ].join('\n'));
  }

  const rolePermissions = resourcePermissionsByRole[resourceRole];

  if (Array.isArray(requiredPermissions)) {
    return allValuesInArray(requiredPermissions, rolePermissions);
  }

  return rolePermissions.includes(requiredPermissions);
}
