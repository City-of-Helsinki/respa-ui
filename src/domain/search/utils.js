import get from 'lodash/get';
import range from 'lodash/range';
import capitalize from 'lodash/capitalize';
import sortBy from 'lodash/sortBy';

import constants from '../../../app/constants/AppConstants';

// TODO: Change to use query-string module.
export const getFiltersFromUrl = (location) => {
  const query = new URLSearchParams(location.search);
  const filters = {};

  query.forEach((value, key) => {
    if (constants.SUPPORTED_SEARCH_FILTERS[key] !== undefined) {
      filters[key] = value;
    }
  });

  return filters;
};

// TODO: Change to use query-string module.
export const getSearchFromFilters = (filters) => {
  const query = new URLSearchParams(filters);
  return query.toString();
};

export const getUnitOptions = (units, locale) => {
  const options = units.map(unit => ({
    value: unit.id,
    label: get(unit, `name[${locale}]`, ''),
  }));

  return sortBy(options, 'label');
};

export const getPurposeOptions = (purposes, locale) => {
  const options = purposes
    .filter(purpose => purpose.parent === null)
    .map(unit => ({
      value: unit.id,
      label: get(unit, `name[${locale}]`, ''),
    }));

  return sortBy(options, 'label');
};

/**
 * Getter for people capacity options.
 * @returns {{label: number, value: number}[]}
 */
export const getPeopleCapacityOptions = () => {
  return [
    ...range(1, 10),
    ...range(10, 35, 5),
    ...range(40, 110, 10),
  ].map(number => ({ label: number, value: number }));
};

/**
 * Getter for municipality options.
 * @returns {{label: string, value: string}[]}
 */
export const getMunicipalityOptions = () => {
  let municipalities = constants.DEFAULT_MUNICIPALITY_OPTIONS;

  if (Array.isArray(SETTINGS.CUSTOM_MUNICIPALITY_OPTIONS)
    && SETTINGS.CUSTOM_MUNICIPALITY_OPTIONS.length) {
    municipalities = SETTINGS.CUSTOM_MUNICIPALITY_OPTIONS;
  }

  return municipalities.map((municipality) => {
    const municipalityStr = typeof municipality === 'string' ? municipality : municipality.toString();

    return {
      value: municipalityStr.toLowerCase(),
      label: capitalize(municipalityStr),
    };
  });
};
