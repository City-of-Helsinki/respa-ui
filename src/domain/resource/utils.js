import round from 'lodash/round';
import filter from 'lodash/filter';
import findIndex from 'lodash/findIndex';
import get from 'lodash/get';
import find from 'lodash/find';
import moment from 'moment';

import * as urlUtils from '../../common/url/utils';
import * as dataUtils from '../../common/data/utils';
import * as reservationUtils from '../reservation/utils';
import constants from '../../../app/constants/AppConstants';
import forEach from "lodash/forEach";


/**
 * Getter for resource page link.
 * Returns a resource page url with the given query params.
 * @param resource
 * @param query
 * @returns {string}
 */
export const getResourcePageLink = (resource, query) => {
  return urlUtils.getLinkString(`/resources/${resource.id}`, query);
};

/**
 * Getter for formatted unit address string.
 * @param unit
 * @param locale
 * @returns {string}
 */
export const getUnitAddress = (unit, locale) => {
  let postal;
  const streetAddress = dataUtils.getLocalizedFieldValue(unit.street_address, locale);
  if (unit && unit.address_zip) {
    postal = unit.address_zip;
  }
  if (unit && unit.municipality) {
    postal = postal ? `${postal} ${unit.municipality}` : unit.municipality;
  }

  return `${streetAddress}${postal ? ` ${postal}` : ''}`;
};

/**
 * Getter for formatted resource distance string.
 * @param resource
 * @returns {string}
 */
export const getResourceDistance = (resource) => {
  const km = resource.distance / 1000;
  return km ? `${round(km, km < 10 ? 1 : null)} km` : '';
};

/**
 * Getter for price string used in resource cards.
 * @param resource
 * @param t
 * @returns {string|*}
 */
export const getPrice = (resource, t) => {
  const minPricePerHour = resource.min_price_per_hour;
  const maxPricePerHour = resource.max_price_per_hour;

  if (!(minPricePerHour || maxPricePerHour)) {
    return t('ResourceIcons.free');
  }

  if (minPricePerHour && maxPricePerHour && minPricePerHour !== maxPricePerHour) {
    return `${Number(minPricePerHour)} - ${Number(maxPricePerHour)} €/h`;
  }

  const priceString = maxPricePerHour || minPricePerHour;
  const price = priceString !== 0 ? Number(priceString) : 0;
  if (price === 0) {
    return t('ResourceIcons.free');
  }

  return price ? `${price} €/h` : null;
};

export const isFree = (resource) => {
  const minPricePerHour = resource.min_price_per_hour;
  const maxPricePerHour = resource.max_price_per_hour;

  if (!(minPricePerHour || maxPricePerHour)) {
    return true;
  }

  if (minPricePerHour && maxPricePerHour && minPricePerHour !== maxPricePerHour) {
    return false;
  }

  const priceString = maxPricePerHour || minPricePerHour;
  const price = priceString !== 0 ? Number(priceString) : 0;

  return price === 0;
};

export const getOpeningHours = (resource, date = null) => {
  if (!resource) {
    return null;
  }

  const openingHours = get(resource, 'opening_hours', []);
  if (!openingHours.length) {
    return null;
  }

  const index = date ? findIndex(openingHours, item => item.date === date) : 0;
  return {
    closes: get(openingHours, `[${index}].closes`),
    opens: get(openingHours, `[${index}].opens`),
  };
};

const reservingIsRestricted = (resource, date) => {
  if (!date) {
    return false;
  }

  const isAdmin = get(resource, 'user_permissions.is_admin', false);
  const isLimited = resource.reservable_before && moment(resource.reservable_before).isBefore(moment(date), 'day');

  return isLimited && !isAdmin;
};

export const getOpenReservations = (resource) => {
  return filter(resource.reservations, (reservation) => {
    return reservation.state !== 'cancelled' && reservation.state !== 'denied';
  });
};

export const getAvailabilityDataForNow = (resource, date) => {
  const openingHours = getOpeningHours(resource, date);
  const reservations = getOpenReservations(resource);

  if (!openingHours || !openingHours.closes || !openingHours.opens) {
    return { status: 'closed', bsStyle: 'danger' };
  }

  const nowMoment = moment();
  const opensMoment = moment(openingHours.opens);
  const closesMoment = moment(openingHours.closes);
  const beginMoment = nowMoment > opensMoment ? nowMoment : opensMoment;

  if (nowMoment > closesMoment) {
    return { status: 'closed', bsStyle: 'danger' };
  }

  const currentReservation = reservationUtils.getCurrentReservation(reservations);
  if (currentReservation || nowMoment < opensMoment) {
    const nextAvailableTime = reservationUtils.getNextAvailableTime(reservations, beginMoment);
    if (nextAvailableTime < closesMoment) {
      return {
        status: 'availableAt',
        bsStyle: 'danger',
        values: { time: nextAvailableTime.format(constants.TIME_FORMAT) },
      };
    }
    return { status: 'reserved', bsStyle: 'danger' };
  }

  return { status: 'available', bsStyle: 'success' };
};

export const getAvailabilityDataForWholeDay = (resource, date) => {
  const openingHours = getOpeningHours(resource, date);

  if (!openingHours || !openingHours.closes || !openingHours.opens) {
    return { status: 'closed', bsStyle: 'danger' };
  }

  if (reservingIsRestricted(resource, date)) {
    return { status: 'reservingRestricted', bsStyle: 'danger' };
  }

  const reservations = getOpenReservations(resource);

  const opensMoment = moment(openingHours.opens);
  const closesMoment = moment(openingHours.closes);
  let total = closesMoment - opensMoment;

  forEach(reservations, (reservation) => {
    const resBeginMoment = moment(reservation.begin);
    if (!resBeginMoment.isSame(opensMoment, 'd')) {
      return;
    }

    const resEndMoment = moment(reservation.end);
    total -= resEndMoment;
    total += resBeginMoment;
  });

  const asHours = moment.duration(total).asHours();
  const rounded = Math.ceil(asHours * 2) / 2;

  if (rounded === 0) {
    return { status: 'reserved', bsStyle: 'danger' };
  }

  return {
    status: 'availableTime',
    bsStyle: 'success',
    values: { hours: rounded },
  };
};
