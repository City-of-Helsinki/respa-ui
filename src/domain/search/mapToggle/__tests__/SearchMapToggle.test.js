import toJSON from 'enzyme-to-json';
import React from 'react';

import SearchMapToggle from '../SearchMapToggle';
import { shallowWithIntl } from '../../../../../app/utils/testUtils';

describe('SearchMapToggle', () => {
  test('renders correctly', () => {
    const wrapper = shallowWithIntl(
      <SearchMapToggle
        active="list"
        onClick={() => null}
        resultCount={53}
      />
    );

    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  test('onClick', () => {
    const onClick = jest.fn();
    const wrapper = shallowWithIntl(
      <SearchMapToggle
        active="list"
        onClick={onClick}
        resultCount={53}
      />
    );

    wrapper.find('.app-SearchMapToggle__button-map').simulate('click');

    expect(onClick.mock.calls[0][0]).toBe('map');
  });
});
