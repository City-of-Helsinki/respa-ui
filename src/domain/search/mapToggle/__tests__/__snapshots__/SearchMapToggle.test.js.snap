// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchMapToggle renders correctly 1`] = `
<div
  className="app-SearchMapToggle"
>
  <Grid
    bsClass="container"
    componentClass="div"
    fluid={false}
  >
    <Row
      bsClass="row"
      componentClass="div"
    >
      <Col
        bsClass="col"
        componentClass="div"
        sm={6}
      >
        <div
          className="app-SearchMapToggle__results-count"
        >
          MapToggle.resultsText
        </div>
      </Col>
      <Col
        bsClass="col"
        componentClass="div"
        sm={6}
      >
        <div
          className="pull-right"
        >
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            className="app-SearchMapToggle__button app-SearchMapToggle__button-list"
            disabled={true}
            key="list"
            onClick={[Function]}
          >
            MapToggle.showList
          </Button>
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            className="app-SearchMapToggle__button app-SearchMapToggle__button-map"
            disabled={false}
            key="map"
            onClick={[Function]}
          >
            MapToggle.showMap
          </Button>
        </div>
      </Col>
    </Row>
  </Grid>
</div>
`;
